slack/plugin.py

```diff
@@ -306,3 +305,106 @@ def slack_conversations_history(
     return messages, next_cursor
+
+
+@purpose("Get conversation replies.")
+def slack_conversation_replies(
+    channel: str,
+    ts: str,
+    cursor: Optional[str] = None,
+    limit: int = 100,
+) -> tuple[list[SlackMessage], str]:
+    """
+    Fetch a page of replies from a specific Slack conversation.
+
+    :param channel: The name or ID of the Slack channel.
+    :param ts: The timestamp of the parent message.
+    :param cursor: Cursor for pagination.
+    :param limit: The maximum number of items to return. May return fewer than the
+        limit, even if there are more items.
+    :return: A tuple containing a list of SlackMessage dataclass instances and the next
+        cursor for pagination. If the next cursor is the empty string, all of the
+        requested items have been returned.
+    """
+    with httpx.Client(
+        transport=AugmentedTransport(actions_v0.authenticated_request_slack_as_user)
+    ) as client:
+        conversation_ids = _conversation_ids_by_name(client)
+        conversation_id = _find_conversation_by_name(conversation_ids, channel)
+        if conversation_id is None:
+            # `channel` does not match any name, so assume that it is an ID.
+            conversation_id = channel
+        params = {
+            "channel": conversation_id,
+            "ts": ts,
+            "limit": limit,
+        }
+        if cursor:
+            params["cursor"] = cursor
+        data = (
+            client.get(
+                "https://slack.com/api/conversations.replies",
+                params=params,
+            )
+            .raise_for_status()
+            .json()
+        )
+    if not data.get("ok", False):
+        if data.get("error") == "channel_not_found":
+            available_channels = f"{sorted(conversation_ids.keys())}"
+            # Avoid making the error message absurdly long.
+            max_length = 1024
+            if len(available_channels) > max_length:
+                truncated = "...(truncated)"
+                available_channels = (
+                    f"{available_channels[:max_length - len(truncated)]}{truncated}"
+                )
+            raise RuntimeError(
+                f"channel '{channel}' not found; "
+                f"available_channels: {available_channels}; "
+                "double-check that you have authorized the correct workspace"
+            )
+        raise RuntimeError(f"fetching replies: {data}")
+    messages = [
+        SlackMessage(
+            type=msg["type"],
+            user=msg.get("user"),
+            text=msg.get("text"),
+            ts=msg["ts"],
+        )
+        for msg in data.get("messages", [])
+    ]
+    next_cursor = data.get("response_metadata", {}).get("next_cursor", "")
+    return messages, next_cursor
+
+
+@purpose("Converts Slack user ID strings to display names")
+def slack_user_lookup(users: Sequence[str]) -> dict[str, SlackUser | None]:
```

Comment:
Are you talking about the input or the output?  If the output, I wanted the API to be able to return information about which slack users don't actually exist (in case the caller is using stale user ids somehow), hence mapping from the user id string to either a real user (with info) or None.

Reply:
I was talking about the input, mostly to signal to callers that it's somewhat nonsensical to have duplicates.  But maybe there is some reason why that is bad: serialization, codegen, whatever...

Reply:
Switched to AbstractSet (should probably use collections.abc.Set but will postpone refactoring to when we are forced to upgrade :)